release/ray_release/reporter/db.py

```diff
@@ -1,18 +1,92 @@
 import time
+import re
 import json
 import boto3
+from typing import List
 from botocore.config import Config
 
 from ray_release.reporter.reporter import Reporter
 from ray_release.result import Result
 from ray_release.config import Test
 from ray_release.logger import logger
 
+CRASH_PATTERN_MAX_LENGTH = 4000
+
 
 class DBReporter(Reporter):
     def __init__(self):
         self.firehose = boto3.client("firehose", config=Config(region_name="us-west-2"))
 
+    def compute_crash_pattern(self, logs: str) -> str:
+        stack_trace = self._compute_stack_trace(logs.splitlines())
+        return self._compute_unique_pattern(stack_trace)[:CRASH_PATTERN_MAX_LENGTH]
+
+    def _compute_unique_pattern(self, stack_trace: List[str]) -> str:
+        """
+        Compute unique pattern from stack trace, by remove factors such as date, time,
+        temp directory, line numbers, etc. This help to aggregate similar logs into
+        same bug patterns
+        """
+        massaged_trace = []
+        for line in stack_trace:
+            line = re.sub(r"\d", "", line.strip())
+            if line == "Traceback (most recent call last):":
+                continue
+            file_line = re.search(r'File "(.*)", (.*)', line)
+            if file_line:
+                line = f'{file_line.group(1).split("/")[-1]}{file_line.group(2)}'
+            massaged_trace.append(line)
+        return "".join(massaged_trace)
+
+    def _compute_stack_trace(self, logs: List[str]) -> List[str]:
+        """
+        Extract stack trace pattern from the logs. Stack trace pattern often matches
+        the following:
+        ERROR ...
+        Traceback (most recent call last):
+            File "...", line ..., in ...
+            ...
+        Exception: exception error
+        """
+        error_stacktrace = []
+        stacktrace = []
+        i = 0
+        while i < len(logs):
```

Comment:
could you add some comments in this function body? the logic is not very easy to read.